File: android/src/main/java/ch/ubique/mapscore/shared/graphics/BitmapTextureHolder.java
package ch.ubique.mapscore.shared.graphics;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.opengl.GLES20;
import android.opengl.GLUtils;

import ch.ubique.mapscore.shared.graphics.objects.TextureHolderInterface;

public class BitmapTextureHolder extends TextureHolderInterface {

	private Bitmap b;
	private int imageWidth;
	private int imageHeight;
	private int textureWidth;
	private int textureHeight;

	public BitmapTextureHolder(Drawable drawable) {
		this(createBitmapFromDrawable(drawable));
	}

	public BitmapTextureHolder(Drawable drawable, int targetWidth, int targetHeight) {
		this(createBitmapFromDrawable(drawable, targetWidth, targetHeight));
	}

	public BitmapTextureHolder(Bitmap bitmap) {
		if (bitmap != null) {
			int width = 2;
			while (width < bitmap.getWidth()) {
				width *= 2;
			}
			int height = 2;
			while (height < bitmap.getHeight()) {
				height *= 2;
			}

			imageWidth = bitmap.getWidth();
			imageHeight = bitmap.getHeight();
			textureWidth = width;
			textureHeight = height;

			if (bitmap.getWidth() != width || bitmap.getHeight() != height) {
				Bitmap large = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
				Canvas c = new Canvas(large);
				c.drawBitmap(bitmap, 0, 0, null);

				//Draw the picture mirrored again to fake clamp mode
				c.save();
				c.scale(1f, -1f, 0, bitmap.getHeight());
				c.drawBitmap(bitmap, 0, 0, null);
				c.restore();
				c.save();
				c.scale(-1f, 1f, bitmap.getWidth(), 0);
				c.drawBitmap(bitmap, 0, 0, null);
				c.restore();
				c.scale(-1f, -1f, bitmap.getWidth(), bitmap.getHeight());
				c.drawBitmap(bitmap, 0, 0, null);

				bitmap.recycle();
				bitmap = large;
			}
		}

		b = bitmap;
	}

	@Override
	public int getImageWidth() {
		return imageWidth;
	}

	@Override
	public int getImageHeight() {
		return imageHeight;
	}

	@Override
	public int getTextureWidth() {
		return textureWidth;
	}

	@Override
	public int getTextureHeight() {
		return textureHeight;
	}

	@Override
	public void attachToGraphics() {
		if (b != null) {
			GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, b, 0);
			b.recycle();
		}
	}

	public Bitmap getBitmap() {
		return b;
	}

	private static Bitmap createBitmapFromDrawable(Drawable drawable) {
		Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(),
				drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
		return createBitmapFromDrawable(drawable, drawable.getIntrinsicWidth(),  drawable.getIntrinsicHeight());
	}

	private static Bitmap createBitmapFromDrawable(Drawable drawable, int targetWidth, int targetHeight) {
		Bitmap bitmap = Bitmap.createBitmap(targetWidth, targetHeight, Bitmap.Config.ARGB_8888);
		Canvas canvas = new Canvas(bitmap);
		drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
		drawable.draw(canvas);
		return bitmap;
	}

}

File: android/src/main/java/io/openmobilemaps/mapscore/shared/graphics/BitmapTextureHolder.kt
/*
 * Copyright (c) 2021 Ubique Innovation AG <https://www.ubique.ch>
 *
 *  This Source Code Form is subject to the terms of the Mozilla Public
 *  License, v. 2.0. If a copy of the MPL was not distributed with this
 *  file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 *  SPDX-License-Identifier: MPL-2.0
 */

package io.openmobilemaps.mapscore.shared.graphics

import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.opengl.GLES20
import android.opengl.GLUtils
import io.openmobilemaps.mapscore.shared.graphics.objects.TextureHolderInterface

class BitmapTextureHolder(bitmap: Bitmap) : TextureHolderInterface() {
	val bitmap: Bitmap
	private var imageWidth = 0
	private var imageHeight = 0
	private var textureWidth = 0
	private var textureHeight = 0

	constructor(drawable: Drawable) : this(createBitmapFromDrawable(drawable)) {}
	constructor(drawable: Drawable, targetWidth: Int, targetHeight: Int) : this(
		createBitmapFromDrawable(drawable, targetWidth, targetHeight)
	) {
	}

	override fun getImageWidth(): Int {
		return imageWidth
	}

	override fun getImageHeight(): Int {
		return imageHeight
	}

	override fun getTextureWidth(): Int {
		return textureWidth
	}

	override fun getTextureHeight(): Int {
		return textureHeight
	}

	override fun attachToGraphics() {
		GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0)
	}

	companion object {
		private fun createBitmapFromDrawable(drawable: Drawable): Bitmap {
			return createBitmapFromDrawable(drawable, drawable.intrinsicWidth, drawable.intrinsicHeight)
		}

		private fun createBitmapFromDrawable(drawable: Drawable, targetWidth: Int, targetHeight: Int): Bitmap {
			val bitmap = Bitmap.createBitmap(targetWidth, targetHeight, Bitmap.Config.ARGB_8888)
			val canvas = Canvas(bitmap)
			drawable.setBounds(0, 0, canvas.width, canvas.height)
			drawable.draw(canvas)
			return bitmap
		}
	}

	init {
		var bitmap = bitmap
		var width = 2
		while (width < bitmap.width) {
			width *= 2
		}
		var height = 2
		while (height < bitmap.height) {
			height *= 2
		}
		imageWidth = bitmap.width
		imageHeight = bitmap.height
		textureWidth = width
		textureHeight = height
		if (bitmap.width != width || bitmap.height != height) {
			val large = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
			val c = Canvas(large)
			c.drawBitmap(bitmap, 0f, 0f, null)

			//Draw the picture mirrored again to fake clamp mode
			c.save()
			c.scale(1f, -1f, 0f, bitmap.height.toFloat())
			c.drawBitmap(bitmap, 0f, 0f, null)
			c.restore()
			c.save()
			c.scale(-1f, 1f, bitmap.width.toFloat(), 0f)
			c.drawBitmap(bitmap, 0f, 0f, null)
			c.restore()
			c.scale(-1f, -1f, bitmap.width.toFloat(), bitmap.height.toFloat())
			c.drawBitmap(bitmap, 0f, 0f, null)
			bitmap.recycle()
			bitmap = large
		}
		this.bitmap = bitmap
	}
}